pkg/scheduler/plugins/reservation/transformer.go

```diff
@@ -22,210 +22,275 @@ import (
-	"github.com/koordinator-sh/koordinator/pkg/util"
 	reservationutil "github.com/koordinator-sh/koordinator/pkg/util/reservation"
 )
 
-// for internal interface testing
-type parallelizeUntilFunc func(ctx context.Context, pieces int, doWorkPiece workqueue.DoWorkPieceFunc)
-
-func defaultParallelizeUntil(handle framework.Handle) parallelizeUntilFunc {
-	return handle.Parallelizer().Until
-}
-
-var (
-	_ frameworkext.PreFilterTransformer = &Plugin{}
-	_ frameworkext.FilterTransformer    = &Plugin{}
-)
-
-func (p *Plugin) BeforePreFilter(handle frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod) (*corev1.Pod, bool) {
-	// skip if the pod is a reserve pod
+func (pl *Plugin) BeforePreFilter(_ frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod) (*corev1.Pod, bool) {
 	if reservationutil.IsReservePod(pod) {
 		return nil, false
 	}
 
-	// list reservations and nodes, and check each available reservation whether it matches the pod or not
-	state, err := p.prepareMatchReservationState(handle, pod)
+	state, err := pl.prepareMatchReservationState(pod)
 	if err != nil {
 		klog.Warningf("BeforePreFilter failed to get matched reservations, err: %v", err)
 		return nil, false
 	}
-	klog.V(4).InfoS("BeforePreFilter successfully prepares reservation state",
-		"pod", klog.KObj(pod))
+	cycleState.Write(stateKey, state)
+
+	klog.V(4).Infof("Pod %v has %d matched reservations and %d unmatched reservations before PreFilter", klog.KObj(pod), len(state.matched), len(state.unmatched))
+	return nil, false
+}
 
-	// initialize a state for matched reservations.
-	// if the pod match any reservations, it stores the matched reservations meta for the pod, mapping from nodeName
-	// to reservationInfo
-	cycleState.Write(preFilterStateKey, state)
+func (pl *Plugin) AfterPreFilter(_ frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod) error {
+	if reservationutil.IsReservePod(pod) {
+		return nil
+	}
 
-	if state.skip {
-		klog.V(5).InfoS("BeforePreFilter skips for no reservation matched", "pod", klog.KObj(pod))
-		return nil, false
+	state := getStateData(cycleState)
+	for nodeName, reservationInfos := range state.unmatched {
+		nodeInfo, err := pl.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
+		if err != nil {
+			continue
+		}
+		if nodeInfo.Node() == nil {
+			continue
+		}
+
+		// Reservations and Pods that consume the Reservations are cumulative in resource accounting.
+		// For example, on a 32C machine, ReservationA reserves 8C, and then PodA uses ReservationA to allocate 4C,
+		// then the record on NodeInfo is that 12C is allocated. But in fact it should be calculated according to 8C,
+		// so we need to return some resources.
+		for _, rInfo := range reservationInfos {
+			if len(rInfo.allocated) == 0 {
+				continue
+			}
+
+			reservePod := reservationutil.NewReservePod(rInfo.reservation)
+			if err := nodeInfo.RemovePod(reservePod); err != nil {
```

Comment:
In unmatched case, the real pod should be deleted, not the reserve pod, as there may be available resources in the reservation.

Reply:
line 85-94 calc the remaining resources and re-add into NodeInfo